arangod/RocksDBEngine/RocksDBDumpManager.cpp

```diff
@@ -0,0 +1,175 @@
+                                                 std::string const& database,
+                                                 std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  auto it = _contexts.find(id);
+  if (it == _contexts.end()) {
+    // "cursor not found" is not a great return code, but it is much more
+    // specific than a generic error. we can also think of a dump context
+    // as a collection of cursors for shard dumping.
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_CURSOR_NOT_FOUND,
+                                   "requested dump context not found");
+  }
+
+  auto context = (*it).second;
+  TRI_ASSERT(context != nullptr);
+  if (!context->canAccess(database, user)) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_FORBIDDEN,
+                                   "insufficient permissions");
+  }
+
+  // whenever we access a context, we proactively extend its lifetime,
+  // so it does not expire quickly after
+  context->extendLifetime();
+
+  return RocksDBDumpContextGuard(*this, std::move(context));
+}
+
+void RocksDBDumpManager::remove(std::string const& id,
+                                std::string const& database,
+                                std::string const& user) {
+  std::lock_guard mutexLocker{_lock};
+
+  auto it = _contexts.find(id);
+  if (it == _contexts.end()) {
+    // "cursor not found" is not a great return code, but it is much more
+    // specific than a generic error. we can also think of a dump context
+    // as a collection of cursors for shard dumping.
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_CURSOR_NOT_FOUND,
+                                   "requested dump context not found");
+  }
+
+  auto& context = (*it).second;
+  if (!context->canAccess(database, user)) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_FORBIDDEN,
+                                   "insufficient permissions");
+  }
+
+  // if we remove the context from the map, then the context will be
+  // destroyed if it is not in use by any other thread. if it is in
+  // use by another thread, the thread will have a shared_ptr of the
+  // context, and the context will be destroyed once the shared_ptr
+  // goes out of scope in the other thread
+  _contexts.erase(it);
+}
+
+void RocksDBDumpManager::dropDatabase(TRI_vocbase_t& vocbase) {
+  std::lock_guard mutexLocker{_lock};
+
+  for (auto it = _contexts.begin(); it != _contexts.end();
+       /* no hoisting */) {
+    auto& context = it->second;
+    if (context->database() == vocbase.name()) {
+      it = _contexts.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+void RocksDBDumpManager::garbageCollect(bool force) {
+  auto const now = TRI_microtime();
+
+  std::lock_guard mutexLocker{_lock};
+
+  for (auto it = _contexts.begin(); it != _contexts.end();
+       /* no hoisting */) {
+    auto& context = it->second;
+
+    if (force || context->expires() < now) {
+      it = _contexts.erase(it);
+    } else {
+      ++it;
+    }
+  }
```

Comment:
This looks like a perfect use case for `std::remove_if`.
```suggestion
  if (force) {
    _contexts.clear();
    return;
  }
  std::remove_if(_contexts.begin(), _contexts.end(), [now](auto& v) { return v.second->expires() < now; }):
  }
```

Reply:
https://stackoverflow.com/a/9210110/3042070

Reply:
Good point! But with C++20 we now have [`erase_if`](https://en.cppreference.com/w/cpp/container/unordered_map/erase_if) which is even nicer since it takes the container instead of the iterators.